support/langkit_support-generic_api.ads

```diff
@@ -0,0 +1,76 @@
+-- terms of the  GNU General Public License  as published by the Free Soft- --
+-- ware Foundation;  either version 3,  or (at your option)  any later ver- --
+-- sion.   This software  is distributed in the hope that it will be useful --
+-- but WITHOUT ANY WARRANTY;  without even the implied warranty of MERCHAN- --
+-- TABILITY  or  FITNESS  FOR A PARTICULAR PURPOSE.                         --
+--                                                                          --
+-- As a special  exception  under  Section 7  of  GPL  version 3,  you are  --
+-- granted additional  permissions described in the  GCC  Runtime  Library  --
+-- Exception, version 3.1, as published by the Free Software Foundation.    --
+--                                                                          --
+-- You should have received a copy of the GNU General Public License and a  --
+-- copy of the GCC Runtime Library Exception along with this program;  see  --
+-- the files COPYING3 and COPYING.RUNTIME respectively.  If not, see        --
+-- <http://www.gnu.org/licenses/>.                                          --
+------------------------------------------------------------------------------
+
+--  This package and its children provide generic APIs so that programs can
+--  work with all Langkit-generated libraries.
+--
+--  Note that it is experimental at this stage, and thus not officially
+--  supported.
+
+limited private with Langkit_Support.Internal;
+
+package Langkit_Support.Generic_API is
+
+   type Language_Id is private;
+   --  Unique identifier for a Langkit-generated library
+
+   function Language_Name (Id : Language_Id) return String;
+   --  Return the name of the language that the library corresponding to ``Id``
+   --  analyzes.
+
+   type Any_Grammar_Rule is new Natural;
+   subtype Grammar_Rule is Any_Grammar_Rule range 1 .. Any_Grammar_Rule'Last;
+   No_Grammar_Rule : constant Any_Grammar_Rule := 0;
+   --  Generic types to designate a grammar rule.
+   --
+   --  A given languages accepts N grammar rules, so the only valid values for
+   --  it are 1 .. N. The Last_Grammar_Rule function below gives the actual N
+   --  for a given language.
+   --
+   --  0 designates the default grammar rule, which is used only in arguments
+   --  passing. The default grammar rule depends on the language, so it is not
+   --  possible to use it directly as default formal value in generic
+   --  subprogram specs.
+
+   function Last_Grammar_Rule (Id : Language_Id) return Grammar_Rule;
+   --  Return the last grammar rule for the given language
+
+   function Default_Grammar_Rule (Id : Language_Id) return Grammar_Rule;
+   --  Return the default grammar rule for the given language
+
+   function Grammar_Rule_Name
+     (Id : Language_Id; Rule : Grammar_Rule) return String;
+   --  Return the camel-with-underscores name for the given grammar rule
+   --  according to the given language. Raise a Precondition_Failure exception
```

Comment:
Missing `` around the exception name. Generally missing those everywhere here :)